File: languages/php/gbdt-par.php
<?php

// gbdt-par.php: parallel GBDT inference via pcntl_fork, sample-band decomposition.
// Invocation: php gbdt-par.php <cores> <n>
// Output: identical to gbdt.php at the same n (core-invariant).
//
// Each worker owns samples [s_start, s_end): evaluates all B trees for its samples,
// returns a packed array of (acc per sample). Parent concatenates bands in index order,
// then runs the serial checksum pass identically to gbdt.php.
//
// The tree arrays (feat, thr, leafval) are read-only and inherited via fork COW.

const GBDT_P          = 1000000007;
const GBDT_D          = 8;
const GBDT_B          = 200;
const GBDT_F          = 8;
const GBDT_NODES      = 511;  // 2^(D+1) - 1
const GBDT_LEAF_START = 255;  // 2^D - 1

// Infer samples [s_start, s_end) and return their acc values.
function infer_band(
    array $feat, array $thr, array $leafval, array $sample,
    int $s_start, int $s_end
): array {
    $B          = GBDT_B;
    $D          = GBDT_D;
    $F          = GBDT_F;
    $NODES      = GBDT_NODES;
    $LEAF_START = GBDT_LEAF_START;

    $accs = [];
    for ($i = $s_start; $i < $s_end; $i++) {
        $sbase = $i * $F;
        $acc   = 0;
        for ($b = 0; $b < $B; $b++) {
            $tbase = $b * $NODES;
            $node  = 0;
            for ($d = 0; $d < $D; $d++) {
                if ($sample[$sbase + $feat[$tbase + $node]] <= $thr[$tbase + $node]) {
                    $node = 2 * $node + 1;
                } else {
                    $node = 2 * $node + 2;
                }
            }
            $acc += $leafval[$tbase + $node];
        }
        $accs[] = $acc;
    }
    return $accs;
}

$cores = isset($argv[1]) ? (int)$argv[1] : 1;
$n     = isset($argv[2]) ? (int)$argv[2] : 5000;

// Build tree arrays (same LCG as serial).
$feat    = array_fill(0, GBDT_B * GBDT_NODES, 0);
$thr     = array_fill(0, GBDT_B * GBDT_NODES, 0);
$leafval = array_fill(0, GBDT_B * GBDT_NODES, 0);

$state = 42;
for ($b = 0; $b < GBDT_B; $b++) {
    $base = $b * GBDT_NODES;
    for ($node = 0; $node < GBDT_LEAF_START; $node++) {
        $state          = ($state * 1103515245 + 12345) & 0x7fffffff;
        $feat[$base + $node] = $state % GBDT_F;
        $state          = ($state * 1103515245 + 12345) & 0x7fffffff;
        $thr[$base + $node]  = $state % 256;
    }
    for ($node = GBDT_LEAF_START; $node < GBDT_NODES; $node++) {
        $state              = ($state * 1103515245 + 12345) & 0x7fffffff;
        $leafval[$base + $node] = $state % 10;
    }
}

$sample = array_fill(0, $n * GBDT_F, 0);
for ($i = 0; $i < $n * GBDT_F; $i++) {
    $state     = ($state * 1103515245 + 12345) & 0x7fffffff;
    $sample[$i] = $state % 256;
}

if ($cores <= 1) {
    // Serial path: identical to gbdt.php
    $t0    = hrtime(true);
    $accs1 = [];
    for ($i = 0; $i < $n; $i++) {
        $sbase = $i * GBDT_F;
        $acc   = 0;
        for ($b = 0; $b < GBDT_B; $b++) {
            $tbase = $b * GBDT_NODES;
            $node  = 0;
            for ($d = 0; $d < GBDT_D; $d++) {
                if ($sample[$sbase + $feat[$tbase + $node]] <= $thr[$tbase + $node]) {
                    $node = 2 * $node + 1;
                } else {
                    $node = 2 * $node + 2;
                }
            }
            $acc += $leafval[$tbase + $node];
        }
        $accs1[] = $acc;
    }
    $ns = hrtime(true) - $t0;
    fwrite(STDERR, "COMPUTE_NS $ns\n");
    $h     = 0;
    $total = 0;
    foreach ($accs1 as $acc) {
        $h     = ($h * 31 + $acc + 1) % GBDT_P;
        $total = ($total + $acc)       % GBDT_P;
    }
    echo $h, "\n";
    echo "gbdt($n) = $total\n";
    exit(0);
}

// Fork workers; each sends back a packed int64 array of acc values for its samples.
$pipes = [];
$pids  = [];

$t0 = hrtime(true);
for ($w = 0; $w < $cores; $w++) {
    $s_start = intdiv($w * $n, $cores);
    $s_end   = intdiv(($w + 1) * $n, $cores);

    $pair = stream_socket_pair(STREAM_PF_UNIX, STREAM_SOCK_STREAM, STREAM_IPPROTO_IP);
    if ($pair === false) {
        fwrite(STDERR, "stream_socket_pair failed\n");
        exit(1);
    }

    $pid = pcntl_fork();
    if ($pid < 0) {
        fwrite(STDERR, "pcntl_fork failed\n");
        exit(1);
    }

    if ($pid === 0) {
        fclose($pair[0]);
        $accs = infer_band($feat, $thr, $leafval, $sample, $s_start, $s_end);
        // Pack as unsigned 64-bit LE; acc is a sum of at most B*9 = 1800, fits in uint64.
        $blob = '';
        foreach ($accs as $v) {
            $blob .= pack('P', $v);
        }
        fwrite($pair[1], $blob);
        fclose($pair[1]);
        exit(0);
    }

    fclose($pair[1]);
    $pipes[$w] = $pair[0];
    $pids[$w]  = $pid;
}

// Collect and concatenate acc arrays in worker (index) order.
$all_accs = array_fill(0, $n, 0);

for ($w = 0; $w < $cores; $w++) {
    $s_start = intdiv($w * $n, $cores);
    $s_end   = intdiv(($w + 1) * $n, $cores);
    $count   = $s_end - $s_start;

    $raw = stream_get_contents($pipes[$w]);
    fclose($pipes[$w]);

    for ($i = 0; $i < $count; $i++) {
        $v = unpack('P', substr($raw, $i * 8, 8));
        $all_accs[$s_start + $i] = $v[1];
    }
}

for ($w = 0; $w < $cores; $w++) {
    pcntl_waitpid($pids[$w], $status);
}
$ns = hrtime(true) - $t0;
fwrite(STDERR, "COMPUTE_NS $ns\n");

// Serial checksum (identical order to gbdt.php).
$h     = 0;
$total = 0;
for ($i = 0; $i < $n; $i++) {
    $acc   = $all_accs[$i];
    $h     = ($h * 31 + $acc + 1) % GBDT_P;
    $total = ($total + $acc)       % GBDT_P;
}
echo $h, "\n";
echo "gbdt($n) = $total\n";

File: languages/swift/gbdt-par.swift
// gbdt-par: parallel sample-band decomposition using DispatchQueue.concurrentPerform.
// Invocation: gbdt-par <cores> <n>
// Output: identical to serial gbdt for any core count (core-invariant).
//
// Tree arrays (feat/thr/leafval) and sample[] are read-only; all workers share them safely.
// Each worker owns a disjoint band of samples and writes accs[i] for its band.
// withUnsafeMutableBufferPointer pins accs[] to avoid COW races.
// Serial checksum iterates accs[] in index order, identical to serial benchmark.
import Foundation

let P_MOD:     Int = 1000000007
let D_TREE:    Int = 8
let B:         Int = 200
let F:         Int = 8
let NODES:     Int = 511       // 2^(D+1) - 1
let LEAF_START: Int = 255      // 2^D - 1

func lcg(_ s: Int) -> Int { return (s &* 1103515245 &+ 12345) & 0x7fffffff }

let args  = CommandLine.arguments
let cores = args.count > 1 ? (Int(args[1]) ?? 1) : 1
let n     = args.count > 2 ? (Int(args[2]) ?? 5000) : 5000

var feat    = [Int32](repeating: 0, count: B * NODES)
var thr     = [Int32](repeating: 0, count: B * NODES)
var leafval = [Int32](repeating: 0, count: B * NODES)

var s = 42
for b in 0..<B {
    let base = b * NODES
    for node in 0..<LEAF_START {
        s = lcg(s); feat[base + node]    = Int32(s % F)
        s = lcg(s); thr [base + node]    = Int32(s % 256)
    }
    for node in LEAF_START..<NODES {
        s = lcg(s); leafval[base + node] = Int32(s % 10)
    }
}

var sample = [Int32](repeating: 0, count: n * F)
for i in 0..<(n * F) { s = lcg(s); sample[i] = Int32(s % 256) }

var accs = [Int](repeating: 0, count: n)

let _t0 = DispatchTime.now().uptimeNanoseconds
feat.withUnsafeBufferPointer { fBuf in
    thr.withUnsafeBufferPointer { tBuf in
        leafval.withUnsafeBufferPointer { lBuf in
            sample.withUnsafeBufferPointer { sBuf in
                accs.withUnsafeMutableBufferPointer { aBuf in
                    DispatchQueue.concurrentPerform(iterations: cores) { w in
                        let sStart = w * n / cores
                        let sEnd   = (w + 1) * n / cores
                        for i in sStart..<sEnd {
                            let sbase = i * F
                            var acc: Int = 0
                            for b in 0..<B {
                                let tbase = b * NODES
                                var node: Int = 0
                                for _ in 0..<D_TREE {
                                    if sBuf[sbase + Int(fBuf[tbase + node])] <= tBuf[tbase + node] {
                                        node = 2 * node + 1
                                    } else {
                                        node = 2 * node + 2
                                    }
                                }
                                acc += Int(lBuf[tbase + node])
                            }
                            aBuf[i] = acc   // disjoint index; no race
                        }
                    }
                }
            }
        }
    }
}
let _ns = DispatchTime.now().uptimeNanoseconds - _t0
fputs("COMPUTE_NS \(_ns)\n", stderr)

// Serial checksum -- same formula and order as serial benchmark.
var h: Int = 0
var total: Int = 0
for i in 0..<n {
    let acc = accs[i]
    h     = (h * 31 + acc + 1) % P_MOD
    total = (total + acc) % P_MOD
}

print(h)
print("gbdt(\(n)) = \(total)")
